textMessages = {'Linear mode' : ['<h3>You are in the <b>Linear mode</b>.</h3>You can point to the particular part of genome and find out its basic information:<ul><li>chromosome [with a helix],</li><li>positions,</li><li>bases and</li><li>SNPs.</li></ul>You can observe this immediately during the surfing space or by calling popup window (left click) with almost the same information and clickable links to NCBI web site for detailed investigations.<br>You can easily increase or decrease the size of the observable area by scrolling the mouse.<br/>Also, you can point to the particular part of genome, adjust observable (<i>linear</i>) area by scrolling, click <b>G</b> key and observe relations between transcription factors and SNPs, which are located under the same sequence in specified plane area of DNA. <h3>Controls</h3><br> <table id = "table_linear"><tr><th>Key</th><th>Description</th></tr><tr><td>W,S,A,D</td><td>Move in space Forward, Back, Left, Right respectively</td></tr><tr><td>Mouse</td><td>Look around</td></tr><tr><td>Left Click</td><td>Get popup window with basic information about chromosome, selected positions, sequence of DNA, and SNPs</td></tr><tr><td>+/- (scroll)</td><td>Change size of observable area.</td></tr><tr><td>Tab</td><td>Get general information about the mode</td></tr><tr><td>M</td><td>Change mode in loop from Linear to Cubic, from Cubic to ChIP-Seq and from ChIP-Seq to Linear mode</td></tr><tr><td>X</td><td>Get popup window with mapped ChIP-Seq data and SNPs</td></tr><tr><td>I</td><td>List all genes available in current cube with some information and opportunity to colour it</td></tr><tr><td>L</td><td>Define the points to measure a distance between them.<br>Show the distance between two already existing points.</td></tr><tr><td>Shift-L</td><td>Erase information about the distance between two existing points (including points themselves). User must use this to measure distance between other two points.</td></tr></table>'],
'Cubic mode' : ['<h3>You are in the <b>Cubic mode</b>.</h3>You can point to the particular part of genome, left click and find out its and neighbour parts of genome\'s basic information:<ul><li>chromosome with a helix,</li><li>positions and</li><li>SNPs.</li></ul>You can easily increase or decrease the size of the observable area by scrolling.<br/>Also, you can point to the particular part of genome, adjust observable area by scrolling, click <b>G</b> key and observe relations between transcription factors and SNPs, which are located under the same sequences in specified cubic area of DNA. <h3>Controls</h3><table id = "table_linear"><tr><th>Key</th><th>Description</th></tr><tr><td>W,S,A,D</td><td>Move in space Forward, Back, Left, Right respectively</td></tr><tr><td>Mouse</td><td>Look around</td></tr><tr><td>+/- (scroll)</td><td>Change size of observable area.</td></tr><tr><td>Left Click</td><td>Get popup window with basic information about selected strands, chromosomes, positions and SNPs</td></tr><tr><td>Tab</td><td>Get general information about the mode</td></tr><tr><td>M</td><td>Change mode in loop from Linear to Cubic, from Cubic to ChIP-Seq and from ChIP-Seq to Linear mode</td></tr><tr><td>X</td><td>Get popup window with mapped ChIP-Seq data and SNPs</td></tr><tr><td>I</td><td>List all genes available in current cube with some information and opportunity to colour it</td></tr><tr><td>L</td><td>Define the points to measure a distance between them.<br>Show the distance between two already existing points.</td></tr><tr><td>Shift-L</td><td>Erase information about the distance between two existing points (including points themselves). User must use this to measure distance between other two points.</td></tr></table>'],
'CHiP-Seq mode' : ['<h3>You are in the <b>CHiP-Seq mode</b>.</h3>All Transcription Factors from ChIP-Seq experiments that are available for this cubic area have been already shown. Colored parts are the proteins that affect parts of genome sequence. According to the level of protein-DNA interaction, green corresponds to the low level(~0) and red one - to the high one(~1000). You can point to the colored part of genome and immediately observe on the bottom screen panel general information about transcription factors, binding site positions and effect level. Left click to the colored parts to observe almost the same information but with links to UCSC Genome Browser for deep investigation.<br><br><b>Attention, if you change the mode you will loose all colored regions.</b><br>This mode is very difficult both for processing and visualization. <h3>Controls</h3> <table id = "table_linear"><tr><th>Key</th><th>Description</th></tr><tr><td>W,S,A,D</td><td>Move in space Forward, Back, Left, Right respectively</td></tr><tr><td>Mouse</td><td>Look around</td></tr><tr><td>Left Click</td><td><b>For colored part of strand:</b> get popup window with basic information about the Transcription Factor, its positions and level of effect</td></tr><tr><td>Tab</td><td>Get general information about the mode</td></tr><tr><td>M</td><td>Change mode in loop from Linear to Cubic, from Cubic to ChIP-Seq and from ChIP-Seq to Linear mode</td></tr><tr><td>T</td><td><b>Colour</b> all strands in current cube according to the ChIP-Seq experiment data: Transcription Factor and effect level (<b>only for cube other from one where mode has been activated</b>).<br><b>Delete</b> stored ChIP-Seq information for current cube (ChIP-Seq data, visualization entities).</td></tr><tr><td>Shift-T</td><td>Refresh field of view: to erase all coloured parts of strands from the memory. (It will immediately improve performance of user\'s local machine.)</td></tr><tr><td>I</td><td>List all genes available in current cube with some information and opportunity to colour it</td></tr><tr><td>L</td><td>Define the points to measure a distance between them.<br>Show the distance between two already existing points.</td></tr><tr><td>Shift-L</td><td>Erase information about the distance between two existing points (including points themselves). User must use this to measure distance between other two points.</td></tr></table>' ]}